From e0a8db025aa34100fc1c74c4c81d24e2fd7693bc Mon Sep 17 00:00:00 2001
From: Frank Meisenbach <fmeisenbach@hilscher.com>
Date: Thu, 18 Mar 2021 23:49:35 +0100
Subject: [PATCH] atheros.c: Fix the vddio-1p8v and smartEEE quirk of Atheros
 ar803x Ethernet-Phy.

Note: These fixes are based on the Linux driver.
---
 drivers/net/phy/atheros.c | 94 +++++++++++++++++++++++++++++++++++++++
 1 file changed, 94 insertions(+)

diff --git a/drivers/net/phy/atheros.c b/drivers/net/phy/atheros.c
index f922fecd6b5..4394ec14235 100644
--- a/drivers/net/phy/atheros.c
+++ b/drivers/net/phy/atheros.c
@@ -12,6 +12,7 @@
 #include <linux/bitfield.h>
 #include <linux/bitops.h>
 #include <dt-bindings/net/qca-ar803x.h>
+#include <malloc.h>
 
 #define AR803x_PHY_DEBUG_ADDR_REG	0x1d
 #define AR803x_PHY_DEBUG_DATA_REG	0x1e
@@ -111,13 +112,90 @@ static int ar803x_enable_tx_delay(struct phy_device *phydev, bool on)
 	return ar803x_debug_reg_mask(phydev, AR803x_DEBUG_REG_5, clear, set);
 }
 
+#define AR803X_DEBUG_REG_31            0x1f
+#define AR803X_VDDIO_1P8V_EN   (1 << 3)
+static inline int ar803x_vddio_1p8v(struct phy_device *phydev, int enable)
+{
+	int ret;
+
+	if (enable)
+		ret = ar803x_debug_reg_mask(phydev, AR803X_DEBUG_REG_31, 0, AR803X_VDDIO_1P8V_EN);
+	else
+		ret = ar803x_debug_reg_mask(phydev, AR803X_DEBUG_REG_31, AR803X_VDDIO_1P8V_EN, 0);
+
+	return ret;
+}
+
+#define AR803X_SMART_EEE_CTRL3	0x805D
+#define AR803X_LPI_EN		(1 << 8)
+static int ar803x_smart_eee(struct phy_device *phydev, int enable)
+{
+	int value;
+	int ret;
+
+	/* 5.1.11 Smart_eee control3 */
+	value = phy_read_mmd(phydev, MDIO_MMD_PCS, AR803X_SMART_EEE_CTRL3);
+
+	(enable) ? (value |= AR803X_LPI_EN) : (value &= ~AR803X_LPI_EN);
+
+	ret = phy_write_mmd(phydev, MDIO_MMD_PCS, AR803X_SMART_EEE_CTRL3, value);
+	if (ret < 0)
+		return ret;
+
+	return 0;
+}
+
+struct at803x_priv {
+	u32 quirks;
+};
+#define AR803X_QUIRKS_SMART_EEE_DISABLE	(1 << 1)
+#define AR803X_QUIRKS_VDDIO_1P8V	(1 << 2)
+
+static int ar803x_quirks(struct phy_device *phydev)
+{
+	struct at803x_priv *priv = phydev->priv;
+	struct ofnode_phandle_args phandle_args;
+
+	if (dev_read_phandle_with_args(phydev->dev, "phy-handle", NULL, 0, 0, &phandle_args)) {
+		debug("Failed to find phy-handle");
+		return -ENODEV;
+	}
+
+	if (ofnode_read_bool(phandle_args.node, "at803x,eee-disabled"))
+		priv->quirks |= AR803X_QUIRKS_SMART_EEE_DISABLE;
+
+	if (ofnode_read_bool(phandle_args.node, "at803x,vddio-1p8v"))
+		priv->quirks |= AR803X_QUIRKS_VDDIO_1P8V;
+
+	return 0;
+}
+
 static int ar8021_config(struct phy_device *phydev)
 {
+	struct at803x_priv *priv = phydev->priv;
+	int ret;
+
+	ret = ar803x_quirks(phydev);
+	if (ret < 0)
+		return ret;
+
 	phy_write(phydev, MDIO_DEVAD_NONE, MII_BMCR,
 		  BMCR_ANENABLE | BMCR_ANRESTART);
 
 	ar803x_enable_tx_delay(phydev, true);
 
+	if (priv->quirks & AR803X_QUIRKS_VDDIO_1P8V) {
+		ret = ar803x_vddio_1p8v(phydev, 1);
+		if (ret < 0)
+			return ret;
+	}
+
+	if (priv->quirks & AR803X_QUIRKS_SMART_EEE_DISABLE) {
+		ret = ar803x_smart_eee(phydev, 0);
+		if (ret < 0)
+			return ret;
+	}
+
 	phydev->supported = phydev->drv->features;
 	return 0;
 }
@@ -335,6 +413,21 @@ static int ar803x_config(struct phy_device *phydev)
 	return 0;
 }
 
+static int ar803x_probe(struct phy_device *phydev)
+{
+	struct at803x_priv *priv;
+
+	priv = malloc(sizeof(*priv));
+	if (!priv)
+		return -ENOMEM;
+
+	memset(priv, 0, sizeof(*priv));
+
+	phydev->priv = priv;
+
+	return 0;
+}
+
 static struct phy_driver AR8021_driver =  {
 	.name = "AR8021",
 	.uid = AR8021_PHY_ID,
@@ -351,6 +444,7 @@ static struct phy_driver AR8031_driver =  {
 	.mask = 0xffffffef,
 	.features = PHY_GBIT_FEATURES,
 	.config = ar803x_config,
+	.probe = ar803x_probe,
 	.startup = genphy_startup,
 	.shutdown = genphy_shutdown,
 };
-- 
2.25.1

